server/channels/app/imaging/orientation.go

```diff
@@ -55,22 +56,77 @@ func MakeImageUpright(img image.Image, orientation int) image.Image {
 	}
 }
 
-// GetImageOrientation reads the input data and returns the EXIF encoded
-// image orientation.
-func GetImageOrientation(input io.Reader) (int, error) {
-	exifData, err := exif.Decode(input)
-	if err != nil {
-		return Upright, fmt.Errorf("failed to decode exif data: %w", err)
+type fwSeeker struct {
+	r io.Reader
+}
+
+func (f *fwSeeker) Read(p []byte) (n int, err error) {
+	return f.r.Read(p)
+}
+
+func (f *fwSeeker) Seek(offset int64, whence int) (int64, error) {
+	// We don't support seeking backwards.
+	if whence != io.SeekCurrent {
+		return 0, fmt.Errorf("unsupported whence value: %d", whence)
 	}
 
-	tag, err := exifData.Get("Orientation")
+	// Seeking forward means we can simply discard the data.
+	n, err := io.CopyN(io.Discard, f.r, offset)
 	if err != nil {
-		return Upright, fmt.Errorf("failed to get orientation field from exif data: %w", err)
+		return n, fmt.Errorf("failed to seek: %w", err)
 	}
 
-	orientation, err := tag.Int(0)
-	if err != nil {
-		return Upright, fmt.Errorf("failed to get value from exif tag: %w", err)
+	return n, nil
+}
+
+// GetImageOrientation reads the input data and returns the EXIF encoded
+// image orientation. Supported formats are JPEG, PNG, TIFF, and WebP.
+// Passing an io.ReadSeeker is preferable as we can't guarantee a plain
+// io.Reader will work for all formats (e.g. TIFF requires backwards seeking).
+func GetImageOrientation(input io.Reader, format string) (int, error) {
+	orientation := Upright
+
+	// Strip the "image/" prefix from the format in case it's a MIME type.
+	format, _ = strings.CutPrefix(format, "image/")
+
+	var imgFormat imagemeta.ImageFormat
+	switch format {
+	case "jpeg":
+		imgFormat = imagemeta.JPEG
+	case "png":
+		imgFormat = imagemeta.PNG
+	case "tiff":
+		imgFormat = imagemeta.TIFF
+	case "webp":
+		imgFormat = imagemeta.WebP
+	default:
+		// We don't support EXIF on any other format.
+		return orientation, fmt.Errorf("unsupported image format: %s", format)
+	}
+
+	var rs io.ReadSeeker
+	if r, ok := input.(io.ReadSeeker); ok {
+		rs = r
+	} else {
+		rs = &fwSeeker{r: input}
+	}
+
+	opts := imagemeta.Options{
+		R: rs,
+		HandleTag: func(tag imagemeta.TagInfo) error {
+			if tag.Tag == "Orientation" {
+				if o, ok := tag.Value.(uint16); ok {
+					orientation = int(o)
```

Comment:
Should we return an error once we've found the orientation, to stop the parsing there? And also implement `ShouldHandleTag` to avoid parsing extra tags before the orientation one?

Reply:
I think we should use `ShouldHandleTag` and then we wouldn't need to stop parsing explicitly. But let me know if you think otherwise.

Reply:
Returning false from `ShouldHandleTag` doesn't stop the parsing, it just skips that's next up, so there's a minor difference. Returning an error from `HandleTag` would stop the parsing completely, so it would save us some CPU cycles. But 0/5, it's not a big difference and if you think `ShouldHandleTag` on its own is much clearer, then we should go with just that.

Reply:
I am all for saving CPU cycles :p But it would mean having to detect the error since I assume it gets forwarded and `Decode` would technically fail. I think it can still be clean enough if we create an error variable. Let me look into it.

Reply:
Done!

Reply:
In case of multiple tags with the same name, we'd pick the first one we find. If the first tag is (for some reason) wrong, the resulting preview may be incorrectly rotated. There's nothing we can do about that as we expect the Exif data to be accurate. Of course these days there are ML vision models that could detect the right orientation just by looking at the content but probably out of scope for now.